Added: Ability to hide subagents from primary agents system prompt.

Sewer56 opened the pull request:
## Summary

Add `hidden` and `permission.task` options to control subagent visibility and invocation permissions.

- fixes #4764
- fixes #4267

## Changes

- Add `hidden: true` option to hide subagents from the `@` autocomplete menu
    - 📓 This is already in `dev` branch
- Add `permission.task` config to control which subagents an agent can invoke via the Task tool (supports allow/deny/ask with glob patterns)
- Filter denied subagents from the Task tool description so the model doesn't attempt to invoke them
- Bypass permission checks when users explicitly invoke subagents via `@` autocomplete

## Purpose

Enables orchestrator-style agent architectures where internal helper subagents should only be invoked programmatically by other agents, not directly by users, while also providing fine-grained control over which agents can call which subagents.

## Modified Files

- `src/agent/agent.ts` - Add `hidden` property and `permission.task` merging logic
- `src/config/config.ts` - Add schema for new config options
- `src/session/prompt.ts` - Track user-invoked agents, regenerate Task tool description with filtered subagents
- `src/tool/task.ts` - Enforce task permissions with user override bypass
- `src/cli/cmd/tui/component/prompt/autocomplete.tsx` - Filter hidden agents from autocomplete
- `test/permission-task.test.ts` - Tests for subagent filtering
- `packages/sdk/js/src/v2/gen/types.gen.ts` - Updated SDK types
- `packages/web/src/content/docs/agents.mdx` - Documentation for new features

Sewer56 committed:
feat: add configurable subagent visibility per agent

- Add subagents config field to agent schema for wildcard-based filtering
- Add filterSubagents helper and runtime validation in Task tool
- Add per-agent subagent filtering in prompt tool resolution
- Add comprehensive tests for subagent filtering patterns

Sewer56 committed:
docs: add subagents configuration documentation

- Document subagents config option for filtering agent invocations
- Add examples for exclusion patterns, wildcards, and precedence rules
- Include Markdown agent configuration example

Sewer56 committed:
fix: TUI autocomplete now respects agent subagents visibility config

- Filter @ autocomplete suggestions based on current agent's subagents configuration
- Fixes bug where all subagents were shown regardless of visibility settings

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.115' into add-hide-subagents

actions-user committed:
chore: format code

malhashemi commented:
I really hope this one gets merged, asked for this to be implemented a while back, thanks for making it happen

Sewer56 committed:
Merge branch 'dev' into add-hide-subagents

Sewer56 commented:
@rekram1-node

actions-user committed:
Update Nix flake.lock and hashes

Sewer56 committed:
Merge branch 'dev' into add-hide-subagents

actions-user committed:
Update Nix flake.lock and hashes

rekram1-node commented:
I think the correct approach here is to allow "task" to have granular permissions like bash does, and then you can whitelist, blacklist, wildcard match, enable/disable subagents

malhashemi commented:
> I think the correct approach here is to allow "task" to have granular permissions like bash does, and then you can whitelist, blacklist, wildcard match, enable/disable subagents 

@rekram1-node I actually have a question here, how does deny vs disable differ for tools/custom tools 

For example with Deny would the agent still see the tool description? The main advantage of restricting subagents to primary is to reduce choice overload and reduce context. So does deny work similarly to disable in that regard?

Sewer56 committed:
Merge branch 'dev' into add-hide-subagents

Sewer56 commented:
@rekram1-node 

I believe what you're thinking of is this:

```jsonc
"agent": {
  "build": {
    "permission": {
      "task": {
        "subagent_name": "deny"
      }
    }
  }
}
```

etc.

However, I chose not to go with that approach; and there is a rather good reason.

The `permission` field semantically speaking declares how the LLM controlling an agent is allowed to execute a tool by default, however the `task` tool is a bit different than the rest of the tools, semantically speaking.

The place where it differs is that unlike others, the `task` tool can be invoked by humans using the `@` notation, with the human being presented a list of subagents to invoke.

What this means is a human could try `@` calling a subagent, and then receive an error saying this is not permitted. You can think of this as typing `!` to execute a command and receiving 'you are not allowed because this command has not been whitelisted in the bash tool'. That would probably be quite weird.

In any case, the original assignment and problem statement in the 2 issues was to *show/hide* items. Although preventing execution of a given subagent is the desired outcome, the other outcome is to avoid discoverability of the subagent by both the LLM by hiding from the system prompt (and the user). If not removed from system prompt, an LLM could still try calling it; be denied, and then get stuck or go off the rails if there are multiple subagents with similar naming. (think `coder` and `coder-high`, etc.)

The current idiom for showing/hiding items within opencode configs is to expose a true/false at the top level; as we have with tools themselves and other parts of the config. There is no precedent for limiting user input/visibility from tool permissions, but there is for top level config items.

------ 

(Had to rewrite this, turns out my reply 10 hours ago didn't send)

malhashemi commented:
@Sewer56 
I do agree with @rekram1-node here in terms of how to disable/enable a subagent for the primary one. Having said that I do agree with you in terms of subagents should not even be mentioned in the primary agent system prompt if they are disabled. I think this goes back to my comment about how does deny vs disable works in opencode. I believe this enable/disable pattern is pretty redundant to permissions. The default behaviour for deny should be to exclude it from the system prompt. This should also not interfere with the human ability to call any subagent through `@` as that should be a seperate invocation pattern. I have not reviewed the parts of the codebase that needs to be touched to achieve this, but from functionality perspective I believe that what would make most sense

rekram1-node commented:
it doesnt have to be that way, if the user is @ it themselves then it could bypass the permission blocks, we already do similar stuff for cwd bypasses

But im not sure what makes the most sense there tbh

rekram1-node commented:
Also if it is set to "deny" then the agent should see the tool as not having that subagent as an option

shuv1337 commented:
For what it's worth, the primary use case of this PR for me (which I have been using in my fork for the last week), is hiding the subagents from the user @ autocomplete list. I have a bunch of small subagents and i restrict them all to a single dedicated primary agent. That way my @ mention list isn't polluted in day to day use with built-in agents, but i can just tab to my 'lazy agent' and then my @ list has all of the subagents there.

Sewer56 commented:
My only question is if you're okay with the semantic inconsistency here.

Technically the permissions field controls (or should control) what an LLM is allowed to invoke.

But in this case, we wouldn't be doing that, the LLM is given free range to call any subagent as it wishes.

When you do an @ call, you ask the LLM to invoke on your behalf, but the LLM is really the one doing the invoking. In that vein, it could re-invoke whenever it wants, or even start guessing names of 'hidden' subagents if they are predictable.

Sure you could do a 'hack', see if the previous user message mentioned a specific subagent and disallow if that was not the case, but even that has some caveats. For instance, the LLM is given free reign to call it 0-* times. Should we only give it one call? Should we give it multiple? There is ambiguity.

-------

In any case.

For me, part of the desired functionality is also to hide it from the humans too. I have subagents that are meant to be purely used by LLMs as part of orchestrator loops.

These subagents aren't used anywhere else but in these orchestrators, and I got multiple variants that use different models for speed/cost tradeoff. Without hiding I would have something like:

```
orchestrator-coder
orchestrator-coder-high
orchestrator-quality-gate-gpt5
orchestrator-quality-gate-opus
orchestrator-commit
orchestrator-planner
orchestrator-searcher-fast
orchestrator-searcher
mcp-search
github
```

every time I type `@`. With no local directories on display, because everything is taken up by the subagents that aren't meant to be invoked by humans.

Sewer56 committed:
Merge branch 'dev' into add-hide-subagents

Sewer56 committed:
Merge branch 'dev' into add-hide-subagents

rekram1-node commented:
I get ur point about no precedent for changing the @ completions, Ig to me whatd make the most sense:

task permission says what agents the agent can access

when you have that agent selected, the @ would respect it…

But ig if you wanted the user to still @ it, it woudlnt be super messy you wouldnt have to read previous user message etc etc, whenever you parse the parts sent by user you would just pass a bypass flag like we do for other part types

I dont know which ux makes more sense tho

Sewer56 commented:
You can send the bypass flag, but as noted above, there is ambiguity of whether you should allow it once, or multiple times, and the LLM could really go with either.

Easy solution is just multiple times, so that flag is essentially sticky.

I don't mind doing that, but then yeah, you still would want the override to hide it to the user.
And the real question is, *where do I place this override*?

There might be users who will still want manual invoke. This wasn't expressed in the issues above, but it's not impossible to imagine.

Technically speaking you could use the perms on the tool to control visibility of subagents to the LLM primary agent. And then use the `subagents` setting introduced here to hide it from the user, or something of the like.

You could even make `subagents` a global thing, but then you need local and global, in case a user wants it in some contexts but not in others. That's where the real ambiguity comes in. Imaginr trying to explain all that in docs too.

I'm open to suggestions based on common consensus. I just originally figured the easiest way forward was to have a simple toggle that flips it for everyone, with no room for confusion. For people who want an escape hatch (e.g. force call a subagent), there would still be option of just swapping primary agent to one that has the subagents available.

malhashemi commented:
> **Disclaimer:** I designed this proposal but used Claude Opus 4.5 to help format and present it clearly.

---

@rekram1-node / @Sewer56 

After thinking through the UX, I'd like to propose an alternative design that separates the two distinct concerns here:

1. **Human visibility**: whether a subagent appears in menus/autocomplete
2. **LLM invocation control**: which subagents a primary agent can spawn

These are independent problems and conflating them creates the semantic confusion discussed above.

---

## Proposed Design

### 1. Human Visibility: `visible` property on subagents

Add a `visible` property (default `true`) that controls whether a subagent appears in the agent selection menu.

**JSON config:**
```json
{
  "$schema": "https://opencode.ai/config.json",
  "agent": {
    "orchestrator-coder": {
      "description": "Internal coding subagent for orchestration loops",
      "mode": "subagent",
      "visible": false
    },
    "orchestrator-planner": {
      "description": "Internal planning subagent",
      "mode": "subagent",
      "visible": false
    },
    "code-reviewer": {
      "description": "Reviews code for best practices",
      "mode": "subagent",
      "visible": true
    }
  }
}
```

**Markdown frontmatter:**
```yaml
---
description: Internal coding subagent for orchestration loops
mode: subagent
visible: false
---

You are a coding subagent. Focus on implementation tasks assigned by the orchestrator.
```

**Rules:**
- `visible: false` hides the subagent from the agent menu.
- Does NOT apply to `mode: primary` or `mode: all` agents (these are always visible and in Tab rotation)
- Does NOT affect whether LLM can invoke the subagent (that's controlled by permissions)

---

### 2. LLM Invocation Control: `permission.task`

Use the existing permissions pattern to control which subagents a primary agent can spawn. This is consistent with how `permission.bash` works.

**JSON config:**
```json
{
  "$schema": "https://opencode.ai/config.json",
  "agent": {
    "build": {
      "mode": "primary",
      "permission": {
        "task": {
          "*": "deny",
          "orchestrator-*": "allow",
          "code-reviewer": "allow"
        }
      }
    },
    "plan": {
      "mode": "primary",
      "permission": {
        "task": {
          "*": "deny"
        }
      }
    },
    "orchestrator": {
      "description": "Main orchestration agent",
      "mode": "primary",
      "permission": {
        "task": {
          "*": "deny",
          "orchestrator-coder": "allow",
          "orchestrator-planner": "allow",
          "orchestrator-quality-gate": "ask"
        }
      }
    }
  }
}
```

**Markdown frontmatter:**
```yaml
---
description: Main orchestration agent
mode: primary
permission:
  task:
    "*": deny
    orchestrator-coder: allow
    orchestrator-planner: allow
    orchestrator-quality-gate: ask
---

You are an orchestrator. Delegate tasks to specialized subagents.
```

**Permission values:**
- `"allow"`: LLM can invoke this subagent freely
- `"ask"`: Prompt user for approval before spawning the subagent
- `"deny"`: LLM cannot invoke this subagent AND it's removed from the Task tool description in the system prompt (preventing confusion/hallucination)

**Wildcard support:**
- `"*": "deny"` — Deny all by default (place first)
- `"orchestrator-*": "allow"` — Allow all subagents starting with `orchestrator-`
- Rules are evaluated sequentially; specific rules override earlier wildcards

---

### 3. Interaction with `tools.task`

For clarity on the hierarchy:

| Config | Effect |
|--------|--------|
| `tools: { task: false }` on primary | Primary cannot spawn ANY subagents (no Task tool) |
| `permission.task.X: "deny"` | Primary cannot spawn subagent X specifically |
| `visible: false` on subagent | Human cannot see in agent menu |

Example - a primary that cannot spawn subagents at all:
```json
{
  "agent": {
    "simple-build": {
      "mode": "primary",
      "tools": {
        "task": false
      }
    }
  }
}
```

---

### 4. TUI Behavior Changes

**When a subagent is active:**
- TUI should display the subagent name (e.g., `[explore]` or `explore`) rather than the primary agent name
- Pressing **Tab** returns to the primary agent
- To select a subagent again, user goes to `/agents` menu or uses keybind

**Agent menu:**
- Shows all primary agents
- Shows all `mode: all` agents
- Shows only subagents where `visible: true` (or not set, since default is `true`)

---

### 5. The `@` Question: Two Options

There are two valid approaches for how humans invoke subagents. I recommend **Option A** but presenting both for discussion:

**Option A (Recommended): Move subagent invocation entirely to agent menu**

- Remove subagents from `@` autocomplete entirely
- `@` is used only for file/context tagging
- Subagents are invoked via `/agents` menu or keybind
- `visible: false` hides from the agent menu

*Rationale:* Subagents are designed to be used by primary agents. If a user needs frequent direct access to a subagent, they should set `mode: all` which makes it a primary agent that's also available as a subagent. This creates a clean separation: `@` = context, agent menu = agents.

*Tradeoff:* More friction for users who want quick inline subagent invocation.

**Option B: Keep `@` for subagents but control autocomplete**

- `@` still invokes subagents
- `visible: false` hides from agent menu.
- Users who know the name can still type `@hidden-subagent` manually

*Rationale:* Preserves quick inline invocation for power users.

*Tradeoff:* Maintains the current inconsistency where `@` is overloaded for both files and agents.

---

### 6. Complete Example

Here's a complete config showing all features:
```json
{
  "$schema": "https://opencode.ai/config.json",
  "agent": {
    "build": {
      "mode": "primary",
      "permission": {
        "task": {
          "*": "deny",
          "code-reviewer": "allow",
          "explore": "allow",
          "general": "allow"
        }
      }
    },
    "orchestrator": {
      "description": "Automated orchestration agent",
      "mode": "primary",
      "permission": {
        "task": {
          "*": "deny",
          "orchestrator-coder": "allow",
          "orchestrator-planner": "allow",
          "orchestrator-quality-gate": "ask"
        }
      }
    },
    "orchestrator-coder": {
      "description": "Internal: Handles coding tasks for orchestrator",
      "mode": "subagent",
      "visible": false
    },
    "orchestrator-planner": {
      "description": "Internal: Handles planning for orchestrator",
      "mode": "subagent",
      "visible": false
    },
    "orchestrator-quality-gate": {
      "description": "Internal: Quality verification",
      "mode": "subagent",
      "visible": false
    },
    "code-reviewer": {
      "description": "Reviews code for best practices and issues",
      "mode": "subagent",
      "visible": true,
      "tools": {
        "write": false,
        "edit": false,
        "task": false
      }
    }
  }
}
```

With this config:
- **Build agent** can invoke `code-reviewer`, `explore`, `general` but NOT any `orchestrator-*` subagents
- **Orchestrator agent** can only invoke its own `orchestrator-*` subagents, with `orchestrator-quality-gate` requiring user approval
- Human sees only `code-reviewer` in the agent menu (plus built-in `explore` and `general`)
- Human does NOT see `orchestrator-coder`, `orchestrator-planner`, or `orchestrator-quality-gate`
- `code-reviewer` does not see the `task` tool.

---

### Summary

| Concern | Config | Location |
|---------|--------|----------|
| Hide subagent from human | `visible: false` | On the subagent |
| Control which subagents LLM can invoke | `permission.task` | On the primary agent |
| Disable Task tool entirely | `tools: { task: false }` | On the primary agent |
| Make subagent also a primary | `mode: all` | On the agent |

This design:
- Separates human UX from LLM behavior
- Is consistent with existing `permission.bash` patterns
- Follows standard config merging (local overrides global)
- Addresses both issues #4764 and #4267

Happy to discuss further or adjust based on feedback.

Sewer56 commented:
This actually works on me, am onboard.

Regarding Option A vs Option B, I got no particular preference.

That said Option A (if chosen) should probably be a follow up PR, since it fundamentally changes how the subagents feature works. I imagine you'd want to ask a broader set of people before changing a feature this ingrained.

Sewer56 commented:
@rekram1-node thoughts? I'm onboard with the proposed approach.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
feat(opencode): add visible field and task permission to agents

design from @malhashemi

malhashemi commented:
Hey @rekram1-node! I noticed @Sewer56 is already working on the proposed design, that's awesome! Could you share your thoughts while there's still work to be done? This feature would really make a big difference in the UX of subagents, so your input is really valued.

Sewer56 commented:
@malhashemi You beat me to it; haha.
Yeah, got first prototype of it here.
I like to review on GitHub so I've pushed it before I've fully reviewed it, not ready yet, but basics seem to be working.

rekram1-node commented:
Okay just to recap I talked w/ dax about this and he said "@" should not hide denied subagents so if the user @s them they should still work, and that the permissions model is what we should use to hide/show etc.

Now the only hiccup here is that we are prolly going to change the name of the task tool to something more "agent-y" so we any change we do here related to it should prolly be experimental for now

rekram1-node commented:
Idk if i like adding this config to hide from user, is that really a pain point for people?

malhashemi commented:
> Idk if i like adding this config to hide from user, is that really a pain point for people?

A massive one for me honestly, imagine having a swarm of subagents not meant to be called by me (I have such in one of the projects) and all they do is clutter file tagging for no reason

rekram1-node commented:
But is it something you are wanting to hide on an agent basis or is it something like hey I never wanna see this appear in my @ results.

Sewer56 commented:
Both.

I'm mostly concerned about hiding from the system prompt personally.

<img width="1074" height="439" alt="image" src="https://github.com/user-attachments/assets/c28b4e20-ce74-4f34-9391-50b7b382da9f" />

Though, admittedly, I wouldn't say this isn't a bit of a nuisance, ahaha.

malhashemi commented:
> But is it something you are wanting to hide on an agent basis or is it something like hey I never wanna see this appear in my @ results.

Agent bases, thats why the agent permission and user tagging is of a different concerns. Plus the 'visible' property does not prevent it from being called, just simply hides it from cluttering file tagging. At least that how I imagine it should work. If a user really wants to he can still call the agent with the `@` but then again this entire thing doesn't make much sense why is agents mixed with files? Agents belong in agents menu.

Probably too ingrained to change now

Sewer56 committed:
fix: don't hide visible==false from system prompt.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
improve: allow user to invoke agents even if they are denied

Sewer56 commented:
In the current state of branch:

- `visible` hides whether it shows up in the autocomplete.
- If an agent is present in user's last message, e.g. `@hidden-agent`, it is permitted to run even if denied.
- If denied by the `task` tool permissions, a subagent is excluded from system prompt; and not allowed to run (unless invoked by user).

There is a caveat. Some models (think Anthropic models with thinking) may be reluctant to try running a subagent that's not in system prompt- thinking it doesn't exist.

We may need to modify the hidden injected text; which currently says `Use the above message and context to generate a prompt and call the task tool with subagent: `

Edit: I added an extra message `. Invoked by user; guaranteed to exist.`, if it's a subagent that's not in system prompt. Seems to work well with Sonnet so far.

Sewer56 committed:
improve: added quick note signifying that the subagent call is user invoked.

Sewer56 commented:
Updated opening post accordingly.

rekram1-node commented:
Okay so dax shipped a refactor that added a new "hidden" field to agents so you can hide from @ completions in your configurations

Sewer56 commented:
(Did it actually ship? I added `hidden: true` to agents on latest dev branch but they still appear)
(Maybe it's in one of the branches like `llm-centralization`)

Edit: https://github.com/sst/opencode/pull/5462
Which means shipped in 1.0.154

Sewer56 commented:
<img width="1756" height="205" alt="image" src="https://github.com/user-attachments/assets/bf2126cb-eb2e-4b5f-b39c-3d9de4e12453" />

But it still appears in .162 , not sure why; code appears fine at glance; it shouldn't show up.

In any case, the code only hides it from autocomplete- I need it removed from the system prompt too FWIW- otherwise the LLM winds up calling the agents when I don't want that.

Sewer56 committed:
refactor: rename visible to hidden for subagent config

Align with dev branch naming convention. The semantic change is:
- visible: false -> hidden: true
- Default behavior unchanged (agents are not hidden by default)

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 commented:
Updated PR to latest `dev`. 
Now with `visible` -> `hidden` renamed to keep track with changes in `dev`- so this extends what dev has got in `llm-centralization` branch.

So now, this PR allows you to hide it from system prompt separately from autocomplete; while still allowing explicit invoke.

Sewer56 commented:
CC @rekram1-node

rekram1-node commented:
Yeah the hidden thing is only visual if you want to hide a subagent from a primary agent, that'd be the purpose of your changes

rekram1-node commented:
/review

rekram1-node commented on packages/opencode/src/tool/task.ts:
```diff
@@ -30,6 +38,31 @@ export const TaskTool = Tool.define("task", async () => {
+                description: params.description,
+              },
+            })
+          }
```
btw don't apply these diffs, typically the agent sucks at diffs but has decent feedback, ill tweak prompt for diffs

Sewer56 commented:
Edits by maintainers is on- if there's anything trivial to change, just commit it directly rather than waiting.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 commented:
Well, what else does it take to get this merged?

I've been very active as far as applying the feedback is concerned; including keeping this branch synced with `dev`; merging conflicts and even adapting changes to it, e.g. renaming `visible` -> `hidden`.

I'm not really sure what I can do here anymore. This is simply a feature I need to save tokens, and prevent the LLM calling agents I don't want called; that simple.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

tusxxx commented:
When it get merged? :-)

malhashemi commented:
The permission system is getting reworked, it is very likely that's what is holding the maintainers from addressing this now.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 commented:
Permission rework started yesterday FWIW with 3f699a91e1f422db5655b222ab453da06c363e36
This PR has been sitting in the current state (since last needed changes to adapt to `dev`) for 7 days :sweat_smile: .
10 if not accounting for renaming `visible` -> `hidden` field for autocomplete visibility.

This will wind up a case where I'll have to rebase/fixup for `dev` branch again. While I don't mind it too much- it is wasted effort, as having it merged *before* the rework started would have avoided conflicts.

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.195' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.197' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 commented:
I didn't know contributing to `OpenCode` was such a difficult job :sweat_smile:

Sewer56 committed:
Merge tag 'v1.0.204' into add-hide-subagents

Sewer56 committed:
Merge remote-tracking branch 'origin/dev' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.220' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.221' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.222' into add-hide-subagents

Sewer56 committed:
Merge tag 'v1.0.223' into add-hide-subagents

Sewer56 committed:
Merge tag 'before-permission-rework' into add-hide-subagents

Sewer56 committed:
feat: allow hiding subagents and controlling task invocation permissions

Sewer56 committed:
Merge 3eb358f: recreate hide subagents feature on permission rework

Force merge to adopt content from 3eb358fe08dcd9cdfff9dd567118ca26e0c17266
while preserving branch history.

Sewer56 committed:
Merge tag 'v1.0.224' into add-hide-subagents

Sewer56 commented:
Recreated the PR to work over the new permissions system.

This is a bit frustrating at this point though.
I don't know what else I need to do to get this crucial feature merged.

I've been keeping this branch very up to date on a daily basis, including resolving merges; and doing the basic testing at each step. I've replied to every comment within a matter of hours (usually less than 4), and I've adapted to every related change made in main, such as introduction of a `hidden` field meaning a rename from `visible` -> `hidden`.

I don't know what else I need to do.

In an ideal world, this would have been merged before work on replacing the permissions system would have started; saving everyone some effort. There was more than a week's time to do this.

However, instead I had to recreate the PR yet again, because the changes were simply too destructive. I'm at a loss- unfortunately.

Sewer56 committed:
fix: treat slash command subagents as user-invoked to bypass permission checks

Sewer56 committed:
fix: treat slash command subagents as user-invoked to bypass permission checks

rekram1-node commented:
@Sewer56 appreciate ur patience and apologies for not being in contact, I will take this across the finish line today/tmr for u so u dont need to slave any longer. 

I try my best to keep up w/ github notifications but across github/x/discord I get hundreds of messages a day and it's really hard to read them all.


My notifications from today:
<img width="98" height="27" alt="Screenshot 2026-01-02 at 3 46 45 PM" src="https://github.com/user-attachments/assets/e53f47fc-a3ce-45d7-a331-d759e62de2a6" />

Sewer56 commented:
<img width="277" height="78" alt="image" src="https://github.com/user-attachments/assets/932fa170-cd8c-48d7-8193-1943f921228d" />

Don't worry- I feel you.
I'm much the same, myself.

A bit less, but I also got users on Discord etc. and other channels to help out.

Sewer56 committed:
Merge tag 'v1.1.1' into add-hide-subagents

malhashemi commented:
@rekram1-node 

I think this is good to go now that v1.1 is released?

Sewer56 commented:
Seems good. I thought I'd need to update docs, but turns out I already did it last time.

Interesting change when you swap out to new `permission` syntax.
With legacy `tools`, the syntax is

```yaml
  task:
    orchestrator-*: allow
    "*": deny
```

But once you swap out to the new `permission` system, you can only use inline JSON.

```yaml
  task: {
    "orchestrator-*": "allow",
    "*": "deny"
  }
```

That caught me offguard for a sec. Seems to be case for other tools too.

Sewer56 committed:
Merge remote-tracking branch 'origin/production' into add-hide-subagents

Sewer56 committed:
fix: include task tool when subagent-specific patterns are allowed

The task tool was incorrectly disabled when using permission rules like:
  task:
    "orchestrator-*": "allow"
    "*": "deny"

The disabled() function was checking evaluate("task", "*", ruleset) which
matched the global deny rule, causing the tool to be removed even though
specific subagent patterns were allowed.

Changes:
- PermissionNext.disabled() now special-cases task tool to check if ANY
  rule is non-deny before disabling
- prompt.ts removes task tool only if no subagents pass the filter
- Fixed pre-existing bugs where evaluate() result (Rule object) was
  compared to string instead of checking .action property

Sewer56 committed:
docs: clarify that last matching rule wins for task permissions

Sewer56 commented:
Hm, caught a subtle bug. When merging latest.

```
  task: {
    "*": "deny",
    "orchestrator-*": "allow"
  }
```

Not sure when it happened, but at some point, this didn't auto activate the `task` tool itself.

So the subagents would be injected, but the `task` tool itself was not added to the list of tools.
I fixed it just now.

It will remove tool if no subagents are runnable. And it'll allow running if at least 1 is runnable. Provided at least 1 rule allows any subagent to run.

Also there was subtle change to `Rule`, where it now has `Rule.action` field. That's also adjusted.

-----

In addition, I noticed in the new permissions system, the rules are applied in declaration order. Meaning that putting `"*": "deny"` as last will in fact deny everything, unlike the legacy system. I adjusted the docs to make a note of this.

All good for merging again.

rekram1-node commented:
this is high on my list, im just out of town for a wedding but ill try to get this in

Edit: just got back from wedding too late tho gotta sleep

rekram1-node committed:
Merge branch 'dev' into add-hide-subagents

rekram1-node committed:
rm todo

malhashemi commented:
Gents with the new permission system, I think we really need this one, right now we can already deny subagents with the the below syntax, but they still clutter the task tool description for no reason:

```
permission:
    task:
        "*": deny
        "codebase-*": allow
```

However, after looking at the code, I think the implementation here could be simplified quite a bit! 

In #7042 (just merged), we exposed `ctx.agent` to tools so they can access the agent's permissions during description generation. We used this to fix skill filtering - the skill tool now filters itself based on permissions.

The same pattern would work beautifully here. Instead of:

1. Building the task description with all agents
2. Exporting TASK_DESCRIPTION and filterSubagents 
3. Having prompt.ts regenerate the description after the fact

We could just do it inside `task.ts` itself:

```ts
export const TaskTool = Tool.define("task", async (ctx) => {
  const agents = await Agent.list().then((x) => x.filter((a) => a.mode !== "primary"))
  
  // Filter by permissions - same pattern as skill.ts
  const agent = ctx?.agent
  const accessibleAgents = agent
    ? agents.filter((a) => {
        const rule = PermissionNext.evaluate("task", a.name, agent.permission)
        return rule.action !== "deny"
      })
    : agents
  const description = DESCRIPTION.replace("{agents}", 
    accessibleAgents.map((a) => `- ${a.name}: ${a.description ?? "..."}`).join("\n")
  )
  
  return { description, parameters, execute }
})
```

This keeps all the filtering logic self-contained in the tool (consistent with how skill.ts works now), avoids the coupling between `task.ts` and `prompt.ts`, and is way fewer lines of code 😄

The `userInvokedAgents` bypass logic might still be needed for the runtime permission check - that part seems fine. But the description filtering could definitely be cleaner!

Happy to help out if you'd like - just let me know!

Sewer56 commented:
I'm happy to get extra hands. Continually updating this to keep up with `dev` for the past 2 months has taken some time alright.

Feel free to drop a patch, or I can just give you write access to my fork, whatever works.

Just remember, that this has to account for the user switching the primary agent. That's why it was done the way it was originally at the time.

malhashemi commented:
> I'm happy to get extra hands. Continually updating this to keep up with `dev` for the past 2 months has taken some time alright.

> Feel free to drop a patch, or I can just give you write access to my fork, whatever works.

If you can grant me write access I would be happy to help

> Just remember, that this has to account for the user switching the primary agent. That's why it was done the way it was originally at the time.

For sure that's the whole value proposition, skill filtration works in the same way.

rekram1-node committed:
Merge branch 'dev' into add-hide-subagents

rekram1-node committed:
clean

rekram1-node commented:
hey today is my first day back I said id merge it :)

Sewer56 commented:
\>w<

Sewer56 commented:
@malhashemi If you wish to, fire away :p
Should have write perms.

rekram1-node commented:
ignore failures, ill patch it and merge

malhashemi commented:
@rekram1-node so do I leave it to you then? just as mentioned we could use the fact that we are exposing the agent now to the tool to make the description filtration self contained within the task tool. Also welcome back I hope you partied hard enough in that wedding, new issues been raining haha

malhashemi commented:
Also to let you know we cannot `@` a subagent if we select a primary that is denying it as of now. Looks like Dax said he doesn't want that behavior but he implemented it himself with that permission PR. As usual keep fixing what Dax breaks :))

Sewer56 commented:
> Also to let you know we cannot @ a subagent if we select a primary that is denying it as of now.

With one of the recent merges from `dev`?
This did work when I tested it 2 days ago with my last push; I wonder if something has changed since.

Edit: Just confirmed.

It should work if the subagent is not hidden. It's one of the things I've had to retest a lot when there have been conflicts with `dev`.

Invokable by user:

<img width="1828" height="814" alt="image" src="https://github.com/user-attachments/assets/6cac4e62-2c8d-441d-ad7d-c3b50db7f82b" />

But not in system prompt:

<img width="1827" height="898" alt="image" src="https://github.com/user-attachments/assets/0bfb5d83-7ca1-4f3a-9522-5aaed1428a91" />

rekram1-node committed:
test: fix

malhashemi commented:
> With one of the recent merges from `dev`? This did work when I tested it 2 days ago with my last push; I wonder if something has changed since.

Oh haven't tested this with your branch yet, but on dev yeah you cannot `@` them now. basically it acts as the primary agent is the one invoking the sub.

Sewer56 commented:
Oh yeah, you can't do that on `dev`, haha.
I had to fix it for this PR, yup.

rekram1-node commented:
I tested quite a bit I think everything works as expected (in this PR)

rekram1-node commented:
waiting for ci to finish then i click the shiny button

malhashemi commented:
So no refactor? If it works do not mess with it haha?

rekram1-node commented:
> So no refactor? If it works do not mess with it haha?

I think the code could be cleaned up a little cause there is already a flag to bypass checks in task tool but the # of lines that arent in test file is like 100 and Idrc its not a huge headache to clean up later

malhashemi commented:
What matters it is working :D thanks @Sewer56 for all the work you put into this, thanks @rekram1-node for merging <3

rekram1-node commented:
yes definitely thanks to @Sewer56 i think they may be a saint, given the amount of patience they had for me xD
